### DISCUSSION TITLE
Disabling Pipeline auto-run from specific git materials

### DISCUSSION THREAD

**Discussion by badbort** [gocd/gocd#9858] [Q&A] [answered]
##### Summary

I need a pipeline that will autorun from one material, but never auto run from a git material. I thought disabling polling from this material would work, but other pipelines 'discover' the commit which triggers the pipeline from this git material.

I don't know if this is a bug with disabled polling. The wording seems to indicate it simply wont check for new commits, but to me that sounds like if a new commit was discovered (somehow, i.e. another pipeline was run) and the pipeline has auto_run enabled, then the pipeline will trigger.

Consider this example:
Git repo Client1
Pipeline A1 produces build artificts from Client1
Pipeline B1 has pipeline A1 as a material, and a git material called PublishScripts. 

I want to trigger A1 after Client1 git repo commits automatically, and B2 should flow on automatically. I want B2 to always use the latest commit of PublishScripts.

My issue is that when this example is expanded with Client2, A2, B2, ect. Committing to PublishScripts is triggering other pipelines.

PublishScripts has polling disabled, as it's a general purpose git repo for maintaining scripts used to publish to our network infrastructure - powershell scripts, ect. Under no circumstances do we want PublishScripts to trigger any pipeline automatically after a commit. Our pipeline should Pipeline B1.. BN should only ever trigger 

Is there any option I am missing? There is an option for disabling this for pipeline materials (ignore_for_scheduling in yaml) yet not for git materials.

**Comment by chadlwilson** [comment, marked as answer]
As to whether it is a bug or not, I don't think I have enough context to comment yet it certainly feels un-intuitive. 

One thing I came across recently in GoCD's own pipelines is leaving the regular polling turned on, but setting a blanket **denylist** of `**/*` on the git material of `PublishScripts`.

I believe the intent, and effect here is to achieve the same thing as you are wanting - so perhaps you can try this as a workaround and see if it gives you the behaviour you expect.

As to the root of the problem, is `PublishScripts` as a material used by any other pipelines where you **do** want it to poll, or where it is configured to do so? Put a different way, if you find the material in the `Materials` view, are there multiple pipelines that show they are being fed from the same material when you click the "Usage" icon? If you look at this pipelines, do they **all** have polling disabled?

My **theory** here is that the way materials are de-duplicated across pipelines perhaps is causing the "should I poll this repo?" setting to be read from a different logical material and overridden during de-duplication. When determining uniqueness, really only the material type, url and branch are included by default. By contrast the allow/deny options are not stored with the material itself, as they belong to its connection to a given pipeline. This challenge with determining "uniqueness of materials" is actually the motivation for the [gocd-git-path-material-plugin](https://github.com/TWChennai/gocd-git-path-material-plugin) since the triggering inconsistency can cause fan-in issues with conflicting allow/denylists for the same repo used by different pipelines.

**Comment by chadlwilson** [comment]
On a side note, thanks for the great description of your problem, and what you are trying to achieve. 👍

**Comment by badbort** [comment]
Thanks for the quick reply!

> One thing I came across recently in GoCD's own pipelines is leaving the regular polling turned on, but setting a blanket denylist of **/* on the git material of PublishScripts.

I might give this a shot and see how it goes after I try reproducing my current issue.

> As to the root of the problem, is PublishScripts as a material used by any other pipelines where you do want it to poll, or where it is configured to do so?

No, all usages are the same in my use case - no client pipelines should ever trigger from PublishScripts at all, and instances of this git material have polling disabled.

After playing around just now, I can reproduce the behaviour easily.

In my earlier example with Client1, pipelines A1, B1, Client 2, A2, B2... ect...

- In a stable gocd environment where nothing is running or scheduled
- Commit changes to PublishScripts
- Observe nothing is scheduled
- Manually run pipeline B2. I have been doing this to fix edge cases or errors.
- B2 is scheduled as you'd expect via manual action
- Observe all other B* pipelines are also scheduled.

```
Triggered by changes
on 09 Nov, 2021 at 15:59:35 Local Time
```

Although polling is disabled, gocd will still 'discover' this new commit after I manually ran B2. This then schedules all dependent pipelines.

I've now added the blacklist filter to all B* pipelines like so:
![image](https://user-images.githubusercontent.com/4551792/140886817-ecf49515-9d2c-4a67-a951-022504c68a11.png)

Sequence of events:
1. Commit a change to PublishScripts
2. Waited for a bit, then manually scheduled B2
3. Observed B2 scheduled as expected. The scheduled pipeline is using the latest commit from PublishScripts, which is desired - good.
4. Stopped B2, and observed no other pipelines were scheduled.

So this blacklist approach is a good workaround, however I think this is a good reason to introduce a new parameter (ignore for scheduling?) which can be configured for each material. Ideally all material types and not just git

The material 'de-duplication' does seem like it might get in the way though...

**Comment by chadlwilson** [comment]
OK, that helps me understand. 

To clarify further, when you earlier said

> Committing to PublishScripts is triggering other pipelines.

..did you actually mean committing, and _then_ triggering a manual pipeline run for 1 of the dependent pipelines? Because if polling is _disabled on all instances_ of that material and there are no webhooks, I can't think of a way committing _on its own_ would trigger anything.

However, by triggering the manual start you are triggering a "manual check" of the material to find its latest revision. At that point GoCD says "well, this wasn't a scheduled _poll_, and I have new information about this repo/branch's latest revision, so I should still check which pipelines are dependent on this material and trigger them if their allow/denylists say it should".

I am also contemplating how this could be documented/displayed differently to make it "feel" more intuitive. I think part of the challenge is the way things are expressed in YAML can give a different impression to the UI, e.g `auto_update` is pretty terse, and doesn't actually document what it means. By contrast the UI displays

![image](https://user-images.githubusercontent.com/29788154/140889695-52855713-ea5d-4254-a2ea-9aa436d73626.png)

**Comment by badbort** [comment]
> ..did you actually mean committing, and then triggering a manual pipeline run for 1 of the dependent pipelines? Because if polling is disabled on all instances of that material and there are no webhooks, I can't think of a way committing on its own would trigger anything.

You are correct, committing did not schedule any pipelines. 

My second post I've tried to observe the exact behaviour and no commits are scheduling anything.

> However, by triggering the manual start you are triggering a "manual check" of the material to find its latest revision. At that point GoCD says "well, this wasn't a scheduled poll, and I have new information about this repo/branch's latest revision, so I should still check which pipelines are dependent on this material and trigger them if their allow/denylists say it should".

Yep, this is the behaviour I am seeing. Manual start triggers a manual check for latest material revision, and then all dependant pipelines are scheduled.

> I am also contemplating how this could be documented/displayed differently to make it "feel" more intuitive

From today experience for me this behaves as I'd expect, but the lack of "ignore for scheduling" option per material adds to the confusion as it feels like a fundamental option that should exist.

**Comment by chadlwilson** [comment]
Yeah, possibly, or some better documentation about how denylists should be used to prevent scheduling/triggering, since it has exactly the same effect as desired.

If you want to have a look, this approach is used on GoCD itself for [its "codesigning" material](https://build.gocd.org/go/materials/value_stream_map/401aecc6c3e989aaa2e4e5004a0aefbc40e726aabca9a889effef0c266dd411e/e11bf602b23df9ba670eb42d336dc6ebfc8204e5) (can login as guest) on stable releases used to sign artifacts, commits etc and provide proof of their origin/authenticity. 

This is basically the same use case as yours. If we make changes to the logic/scripts/code for how we sign/produce new releases we don't want to actually trigger and produce a new stable release with all its signed artifacts. However the same material is also used earlier in the pipeline to produce experimental/trial releases on every tested commit, and in these cases we do want it to trigger and validate the signing logic using the latest scripting/code. Denylists are used throughout the "produce stable release" pipelines to prevent these earlier runs with "latest codesigning scripting" triggering a re-run of a stable release.



![image](https://user-images.githubusercontent.com/29788154/140894895-e49ac039-04de-411a-8561-5bdf33b436c1.png)

**Comment by badbort** [comment]
> Yeah, possibly, or some better documentation about how denylists should be used to prevent scheduling/triggering, since it has exactly the same effect as desired.

For pipeline materials there appears to be this option which my interpretation is the desired option I'd want on the git material:
![image](https://user-images.githubusercontent.com/4551792/140906170-fba6046b-e079-468e-aec6-c784e2a4297c.png)

https://github.com/gocd/gocd/blob/b80fb11c56f1135c705617a88f4da167d0436517/domain/src/main/java/com/thoughtworks/go/config/materials/AbstractMaterial.java#L239

Perhaps it's worth implementing [ignoreForScheduling](https://github.com/gocd/gocd/blob/b80fb11c56f1135c705617a88f4da167d0436517/domain/src/main/java/com/thoughtworks/go/config/materials/AbstractMaterial.java#L239) and having this functionality standardized across all material types. Yaml pipelines will also be less confusing with a common property

**Comment by badbort** [comment]
Thanks for your help. Should I close this?

**Comment by chadlwilson** [comment]
I converted this to a discussion, and created a summary issue at #9859 to help discovery if others have the same issue.
